### ISSUE TITLE
Error: dh key too small

### ISSUE THREAD

**Issue by rmendal** [hootnot/oanda-api-v20#163]
I've got an issue where my app will run fine on either of my laptops but fails on my raspberry pi's. Below I list the errors and the relevant info on both pis.

First pi runs Ubuntu 20.04

> Traceback (most recent call last):                                                             
  File "/usr/lib/python3/dist-packages/urllib3/contrib/pyopenssl.py", line 453, in wrap_socket
    cnx.do_handshake()            
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1915, in do_handshake
    self._raise_ssl_error(self._ssl, result)
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1647, in _raise_ssl_error
    _raise_current_error()                  
  File "/usr/lib/python3/dist-packages/OpenSSL/_util.py", line 54, in exception_from_error_queue
    raise exception_type(errors)    
OpenSSL.SSL.Error: [('SSL routines', 'tls_process_ske_dhe', 'dh key too small')]
                                               
>During handling of the above exception, another exception occurred:                                             
 Traceback (most recent call last):                                                                                                                                                            
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)                                                                                                                                                                          
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 343, in _make_request
    self._validate_conn(conn)                                                                                                                                                                 
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 841, in _validate_conn
    conn.connect()                                                                             
  File "/usr/lib/python3/dist-packages/urllib3/connection.py", line 344, in connect
    ssl_context=context)                                                                       
  File "/usr/lib/python3/dist-packages/urllib3/util/ssl_.py", line 344, in ssl_wrap_socket
    return context.wrap_socket(sock, server_hostname=server_hostname)          
  File "/usr/lib/python3/dist-packages/urllib3/contrib/pyopenssl.py", line 459, in wrap_socket
    raise ssl.SSLError('bad handshake: %r' % e)                                
ssl.SSLError: ("bad handshake: Error([('SSL routines', 'tls_process_ske_dhe', 'dh key too small')])",)

>During handling of the above exception, another exception occurred:                                                                                                                                                                                              
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 449, in send
    timeout=timeout
  File "/usr/lib/python3/dist-packages/urllib3/connectionpool.py", line 638, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/usr/lib/python3/dist-packages/urllib3/util/retry.py", line 398, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='api-fxpractice.oanda.com', port=443): Max retries exceeded with url: /v3/instruments/EUR_USD/candles?count=50&granularity=H4 (Caus
ed by SSLError(SSLError("bad handshake: Error([('SSL routines', 'tls_process_ske_dhe', 'dh key too small')])")))

>During handling of the above exception, another exception occurred:
Traceback (most recent call last):
  File "pyfor_funcs.py", line 10, in <module>
    from strategy import Strategies
  File "/home/pi/pyfor/MA-Specific/strategy.py", line 60, in <module>
    s = Strategies(c.get_candles(), u.count)
  File "/home/pi/pyfor/MA-Specific/candles.py", line 51, in get_candles
    close_list.append(self.close(x))
  File "/home/pi/pyfor/MA-Specific/candles.py", line 46, in close
    return self.ohlc(data)[3]
  File "/home/pi/pyfor/MA-Specific/candles.py", line 24, in ohlc
    UserData.client.request(UserData.o)
  File "/home/pi/.local/lib/python3.7/site-packages/oandapyV20/oandapyV20.py", line 306, in request
    request_args, headers=headers)
  File "/home/pi/.local/lib/python3.7/site-packages/oandapyV20/oandapyV20.py", line 234, in __request
    raise err
  File "/home/pi/.local/lib/python3.7/site-packages/oandapyV20/oandapyV20.py", line 231, in __request
    **request_args)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 546, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 533, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 646, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 514, in send
    raise SSLError(e, request=request)
requests.exceptions.SSLError: HTTPSConnectionPool(host='api-fxpractice.oanda.com', port=443): Max retries exceeded with url: /v3/instruments/EUR_USD/candles?count=50&granularity=H4 (Caused b
y SSLError(SSLError("bad handshake: Error([('SSL routines', 'tls_process_ske_dhe', 'dh key too small')])")))

Here is some relevant system info on the pi...
>ubuntu@ubuntu:~$ python3 -m OpenSSL.debug
pyOpenSSL: 19.1.0
cryptography: 2.9.2
cffi: 1.14.0
cryptography's compiled against OpenSSL: OpenSSL 1.1.1f  31 Mar 2020
cryptography's linked OpenSSL: OpenSSL 1.1.1f  31 Mar 2020
Pythons's OpenSSL: OpenSSL 1.1.1f  31 Mar 2020
Python executable: /usr/bin/python3
Python version: 3.8.2 (default, Apr 27 2020, 15:53:34) 
[GCC 9.3.0]
Platform: linux
sys.path: ['/home/ubuntu', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '/home/ubuntu/.local/lib/python3.8/site-packages', '/usr/local/lib/python3.8/dist-packages', '/usr/lib/python3/dist-packages']

Pip Freeze...
>astroid==2.4.1                                                                                                                                                                        [58/104]
attrs==19.3.0           
Automat==0.8.0   
blinker==1.4       
certifi==2019.11.28
cffi==1.14.0        
chardet==3.0.4
Click==7.0           
cloud-init==20.1 
colorama==0.4.3
command-not-found==0.3      
configobj==5.0.6
constantly==15.1.0      
convertdate==2.2.1      
cryptography==2.9.2
cutie==0.2.2  
cycler==0.10.0       
Cython==0.29.17
dbus-python==1.2.16  
distro==1.4.0    
distro-info===0.23ubuntu1
entrypoints==0.3
fail2ban==0.11.1
holidays==0.10.2
httplib2==0.14.0
hyperlink==19.0.0
idna==2.8
importlib-metadata==1.5.0
incremental==16.10.1
isort==4.3.21
Jinja2==2.10.1
jsonpatch==1.22
jsonpointer==2.0
jsonschema==3.2.0
keyring==18.0.1
kiwisolver==1.2.0
korean-lunar-calendar==0.2.1
language-selector==0.1
launchpadlib==1.10.13
lazr.restfulclient==0.14.2
lazr.uri==1.0.3
lazy-object-proxy==1.4.3
linecache2==1.0.0                                                                                                                                                                     
lxml==4.5.0
MarkupSafe==1.1.0
matplotlib==3.2.1
mccabe==0.6.1
more-itertools==4.2.0
netifaces==0.10.4
numpy==1.18.4
oandapyV20==0.6.3
oauthlib==3.1.0
pandas==1.0.3
pandas-datareader==0.8.1
pbr==5.4.5
pyasn1==0.4.2
pyasn1-modules==0.2.1
pycparser==2.20
pycrypto==2.6.1
PyGObject==3.36.0
PyHamcrest==1.9.0
pyinotify==0.9.6
PyJWT==1.7.1
pylint==2.5.0
pymacaroons==0.13.0
PyMeeus==0.3.7
PyNaCl==1.3.0
pyOpenSSL==19.1.0
pyparsing==2.4.7
pyrsistent==0.15.5
pyserial==3.4
python-apt==2.0.0
python-dateutil==2.8.1
python-debian===0.1.36ubuntu1
pytz==2019.3
PyYAML==5.3.1
readchar==2.0.1
requests==2.22.0
requests-unixsocket==0.2.0
SecretStorage==2.3.1
service-identity==18.1.0
simplejson==3.16.0
six==1.14.0
ssh-import-id==5.10
systemd-python==234
testresources==2.0.0
toml==0.10.0
traceback2==1.4.0
tulipy==0.4.0
Twisted==18.9.0
ubuntu-advantage-tools==20.3
ufw==0.36
unattended-upgrades==0.1
unittest2==1.1.0
urllib3==1.25.8
wadllib==1.3.3
wrapt==1.12.1
zipp==1.0.0
zope.interface==4.7.1

>ubuntu@ubuntu:~$ python3 --version
Python 3.8.2

----------------------------------------------------------------
Second pi runs raspbian stretch

>pi@raspi-dev:~ $ python3 -m OpenSSL.debug
pyOpenSSL: 19.0.0
cryptography: 2.6.1
cffi: 1.14.0
cryptography's compiled against OpenSSL: OpenSSL 1.1.1d  10 Sep 2019
cryptography's linked OpenSSL: OpenSSL 1.1.1d  10 Sep 2019
Pythons's OpenSSL: OpenSSL 1.1.1d  10 Sep 2019
Python executable: /usr/bin/python3
Python version: 3.7.3 (default, Dec 20 2019, 18:57:59) 
[GCC 8.3.0]
Platform: linux
sys.path: ['/home/pi', '/usr/lib/python37.zip', '/usr/lib/python3.7', '/usr/lib/python3.7/lib-dynload', '/home/pi/.local/lib/python3.7/site-packages', '/usr/local/lib/python3.7/dist-packages', '/usr/lib/python3/dist-packages']

pip freeze

asn1crypto==0.24.0                                                                                                                                                                  
astroid==2.4.1  
certifi==2018.8.24  
cffi==1.14.0
chardet==3.0.4    
colorama==0.4.3    
convertdate==2.2.1
cryptography==2.6.1
cutie==0.2.2    
cycler==0.10.0 
Cython==0.29.17
entrypoints==0.3 
fail2ban==0.10.2
holidays==0.10.2
idna==2.6
isort==4.3.21
keyring==17.1.1
keyrings.alt==3.1.1
kiwisolver==1.2.0
korean-lunar-calendar==0.2.1
lazy-object-proxy==1.4.3
lxml==4.5.0
matplotlib==3.2.1
mccabe==0.6.1
numpy==1.18.4
oandapyV20==0.6.3
pandas==1.0.3
pandas-datareader==0.8.1
pycparser==2.20
pycrypto==2.6.1
PyGObject==3.30.4
pyinotify==0.9.6
pylint==2.5.0
PyMeeus==0.3.7
pyOpenSSL==19.0.0
pyparsing==2.4.7
python-apt==1.8.4.1
python-dateutil==2.8.1
pytz==2019.3
pyxdg==0.25
readchar==2.0.1
requests==2.21.0
SecretStorage==2.3.1
six==1.12.0   
ssh-import-id==5.7
systemd-python==234
toml==0.10.0       
tulipy==0.4.0         
typed-ast==1.4.1
urllib3==1.24.1
wrapt==1.12.1  

pi@raspi-dev:~ $ python3 --version
Python 3.7.3

Same errors on both pis. Again, I don't get these errors on my laptops which run the same output as the first pi. The app runs normally and I can query info form the API, place orders, pull candles and so on.

**Comment by rmendal** [comment]
Based on someone else having this same issue but in a different way with OandaAPI and python I don't believe this is actually a problem with the wrapper.

**Comment by hootnot** [comment]
No it is not a problem in oandapyV20.
I recently ran into this same error deploying _oandapyV20_ in a docker container. That container had no _openssl_.conf. Making that available the problem as fixed.

So you need to seek the solution in your SSL-config. Seeking for a solution of my problem I ran into: https://stackoverflow.com/questions/38015537/python-requests-exceptions-sslerror-dh-key-too-small . It suggests to comment the line: _CipherString_ but I did not even have that line in my config.

**Comment by rmendal** [comment]
> No it is not a problem in oandapyV20.
> I recently ran into this same error deploying _oandapyV20_ in a docker container. That container had no _openssl_.conf. Making that available the problem as fixed.
> 
> So you need to seek the solution in your SSL-config. Seeking for a solution of my problem I ran into: https://stackoverflow.com/questions/38015537/python-requests-exceptions-sslerror-dh-key-too-small . It suggests to comment the line: _CipherString_ but I did not even have that line in my config.

I've done this and tried DEFAULTSECLEVEL=2  and 1. It did not fix the issue

**Comment by hootnot** [comment]
and commenting that line out as the stackoverflow solution suggests ?

**Comment by rmendal** [comment]
> and commenting that line out as the stackoverflow solution suggests ?

The line didn't exist initially so I added it. It didn't work before and it doesn't work now. I've found several other related questions on stack that had various suggestions to add it, change the level, remove it.

I've got the same issue across different versions of python (3.7.3 vs 3.8.3), openssl (1.1.1d vs 1.1.1f), and pip packages. I've tried upgrading all my pip packages to the latest versions and still nothing.

Again, this works on my laptops without issue. I'm not sure what I'm missing.
